File: pibridge/src/agent-groups.js
// Agent Groups: a roster of specialist models sitting next to the orchestrator.
//
// The technician talks to ONE model (the orchestrator). That model can `delegate`
// a narrow job to a cheaper specialist with a fresh context window, so a long
// ticket does not keep re-sending every file read / log dump through Grok.
//
// Roles with a workspace (Coding group, optional path) get real file tools in an
// isolated child session. Roles without a workspace run as a think-only one-shot
// (scout a blob of text, draft a plan, review a script, summarise). The parent
// keeps the TRMM tools.

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import {
  createAgentSession,
  DefaultResourceLoader,
  SessionManager,
} from "@earendil-works/pi-coding-agent";
import { Type } from "typebox";
import { registerModels } from "./models-catalog.js";

const ROLE_TOOLS = {
  scout: ["read", "grep", "find", "ls"],
  files: ["read", "ls"],
  grep: ["grep", "find"],
  planner: ["read", "grep", "find", "ls"],
  coder: ["read", "grep", "find", "ls", "bash", "edit", "write"],
  reviewer: ["read", "grep", "find", "ls"],
};

const THINK_ROLES = new Set(["orchestrator", "operator", "summarizer", "planner", "reviewer", "scout"]);

export function makeGroupState(blob) {
  return { current: blob?.agent_group || null, rt: null };
}

export function mergeGroupKeys(keys, blob) {
  const extra = blob?.agent_group_keys || {};
  for (const [p, k] of Object.entries(extra)) if (k) keys[p] = k;
  return keys;
}

export async function ensureGroupModels(rt, blob, log) {
  const group = blob?.agent_group;
  const members = group?.members || group?.roles || [];
  if (!members.length) return;
  const wanted = [];
  for (const m of members) {
    if (!m.provider || !m.model_id) continue;
    if (rt.findModel(m.provider, m.model_id)) continue;
    wanted.push({
      provider: m.provider,
      model_id: m.model_id,
      display_name: m.display_name || m.model_id,
    });
  }
  if (!wanted.length) return;
  const providers = Object.entries(blob.agent_group_keys || {}).map(([name, api_key]) => ({
    name,
    api_key,
  }));
  try {
    const out = await registerModels(providers, rt, wanted);
    log?.("agent_group_register", wanted.map((w) => `${w.provider}/${w.model_id}`).join(","), out?.registered || []);
  } catch (e) {
    log?.("agent_group_register_error", String(e?.message || e).slice(0, 200));
  }
}

export function groupPromptAppendix(group) {
  if (!group) return "";
  const members = group.members || group.roles || [];
  if (!members.length) return "";
  const lines = members.map((m) => {
    const name = m.display_name || m.model_id;
    const def = String(m.definition || "").trim().split(/\n/)[0] || "";
    return def
      ? `- ${m.role}: ${name} (${m.provider}/${m.model_id}) — ${def}`
      : `- ${m.role}: ${name} (${m.provider}/${m.model_id})`;
  });
  return (
    `\n\nAGENT GROUP: ${group.name}` +
    (group.description ? `\n${group.description}` : "") +
    `\nYou are the ORCHESTRATOR. The technician talks only to you. You have a team:\n` +
    lines.join("\n") +
    `\n\nHOW TO USE THEM (this is how we keep cost down):` +
    `\n- Call the \`delegate\` tool for recon, search, planning, review, or a first-pass summary.` +
    `\n- Each delegate starts with a CLEAN context. Give it everything it needs in \`task\`.` +
    `\n- Do NOT dump raw logs, whole files, or huge tool results into THIS conversation if a specialist can compress them first.` +
    `\n- You still run the TRMM / ticket tools yourself. Specialists think or read; they do not mutate devices.` +
    `\n- Prefer scout → planner → (you apply the plan) over doing the recon yourself.` +
    (group.workspace ? `\n- File/grep/coder subagents run in workspace: ${group.workspace}` : "") +
    `\n`
  );
}

export function attachGroupToLoader(loaderOpts, groupState) {
  const baseOverride = loaderOpts.systemPromptOverride;
  return {
    ...loaderOpts,
    extensionFactories: [
      ...(loaderOpts.extensionFactories || []),
      (pi) => registerDelegateTool(pi, groupState),
    ],
    systemPromptOverride: () => {
      const base = typeof baseOverride === "function" ? baseOverride() : (baseOverride || "");
      return String(base || "") + groupPromptAppendix(groupState.current);
    },
  };
}

function registerDelegateTool(pi, groupState) {
  pi.registerTool({
    name: "delegate",
    label: "Delegate",
    description:
      "Hand a narrow job to a specialist in this agent group. The specialist has a " +
      "fresh context window (it does NOT see this conversation). Use scout/grep/files " +
      "for recon, planner for a plan, reviewer for a review, summarizer to compress a " +
      "blob of text, coder to write code (only when a workspace is configured), " +
      "operator to draft a device fix. You keep the TRMM tools.",
    parameters: Type.Object({
      role: Type.String({
        description: "Specialist role: scout, files, grep, planner, coder, operator, reviewer, summarizer",
      }),
      task: Type.String({
        description: "Self-contained instructions. The specialist cannot see this chat.",
      }),
    }),
    async execute(_toolCallId, params, signal) {
      const group = groupState.current;
      if (!group) {
        return {
          content: [{ type: "text", text: "No agent group is active. Work the task yourself." }],
          details: {},
        };
      }
      const role = String(params.role || "").trim().toLowerCase();
      const task = String(params.task || "").trim();
      if (!task) {
        return { content: [{ type: "text", text: "delegate requires a task." }], details: {} };
      }
      if (role === "orchestrator") {
        return {
          content: [{ type: "text", text: "You ARE the orchestrator. Do not delegate to yourself." }],
          details: {},
        };
      }
      const member = (group.members || group.roles || []).find((m) => m.role === role && m.enabled !== false);
      if (!member) {
        const have = (group.members || group.roles || []).map((m) => m.role).join(", ");
        return {
          content: [{ type: "text", text: `No '${role}' specialist in group ${group.name}. Available: ${have}` }],
          details: {},
        };
      }
      try {
        const text = await runSpecialist({ group, member, task, signal, rt: groupState.rt });
        return {
          content: [{ type: "text", text: text || "(specialist returned nothing)" }],
          details: { role, model: `${member.provider}/${member.model_id}` },
        };
      } catch (e) {
        return {
          content: [{ type: "text", text: `Specialist ${role} failed: ${String(e?.message || e).slice(0, 400)}` }],
          details: { role, error: true },
        };
      }
    },
  });
}

async function runSpecialist({ group, member, task, signal, rt }) {
  if (!rt) throw new Error("agent-group runtime is not ready");
  let model = rt.findModel(member.provider, member.model_id);
  if (!model) {
    await registerModels(
      [{ name: member.provider, api_key: "" }],
      rt,
      [{ provider: member.provider, model_id: member.model_id, display_name: member.display_name || member.model_id }],
    );
    model = rt.findModel(member.provider, member.model_id);
  }
  if (!model) {
    throw new Error(`model ${member.provider}/${member.model_id} is not available to the runtime`);
  }

  const workspace = String(group.workspace || "").trim();
  const wantTools = ROLE_TOOLS[member.role];
  const useFiles = !!(workspace && wantTools && fs.existsSync(workspace));
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), `pi-group-${member.role}-`));
  const cwd = useFiles ? workspace : tmp;
  const system = specialistPrompt(member, group, useFiles);

  const loader = new DefaultResourceLoader({
    agentDir: tmp,
    cwd,
    noExtensions: true,
    systemPromptOverride: () => system,
  });
  await loader.reload();
  const sessionManager = SessionManager.create(tmp);
  const opts = {
    model,
    thinkingLevel: member.thinking_level || "medium",
    ...rt.sessionOpts,
    resourceLoader: loader,
    sessionManager,
    agentDir: tmp,
    cwd,
  };
  if (useFiles) opts.tools = wantTools;
  else opts.noTools = "all";

  const { session } = await createAgentSession(opts);
  if (signal?.aborted) throw new Error("aborted");
  const abort = () => { try { session.abort?.(); } catch { /* noop */ } };
  signal?.addEventListener?.("abort", abort, { once: true });
  try {
    await session.prompt(task);
  } finally {
    signal?.removeEventListener?.("abort", abort);
    try { session.dispose?.(); } catch { /* noop */ }
  }
  const text = lastAssistantText(session);
  try { fs.rmSync(tmp, { recursive: true, force: true }); } catch { /* noop */ }
  return text;
}

function lastAssistantText(session) {
  const msgs = session?.messages || [];
  for (let i = msgs.length - 1; i >= 0; i--) {
    const m = msgs[i];
    if (m?.role !== "assistant") continue;
    const parts = Array.isArray(m.content) ? m.content : [];
    const text = parts.filter((p) => p.type === "text").map((p) => p.text).join("");
    if (text) return text;
  }
  return "";
}

function specialistPrompt(member, group, useFiles) {
  const role = member.role;
  const common =
    `You are the ${role} specialist in the "${group.name}" agent group.\n` +
    `You do not talk to the technician. Return a compact, useful result to the orchestrator.\n` +
    `Do not greet. Do not offer next steps unless asked. No fluff.\n`;
  const written = String(member.definition || "").trim();
  if (written) {
    return common + written + (useFiles ? "\nA workspace is configured; you may read and (if this role writes) edit files there." : "");
  }
  return common + "Do the assigned task and return the result.";
}

export function findGroupInBlob(blob, groupId) {
  if (groupId == null || groupId === "" || groupId === 0) return null;
  const id = Number(groupId);
  return (blob.agent_groups || []).find((g) => Number(g.id) === id) || null;
}

export function summarizerMember(group) {
  if (!group) return null;
  return (group.members || group.roles || []).find((m) => m.role === "summarizer" && m.enabled !== false) || null;
}

export function publicReady(group) {
  if (!group) return null;
  return {
    id: group.id,
    name: group.name,
    slug: group.slug,
    kind: group.kind,
    is_default: !!group.is_default,
  };
}

// silence unused in case a role catalog is referenced later
void THINK_ROLES;

File: api/tacticalrmm/core/agent_groups.py
"""Agent Groups: a named team of specialist models for one Pi chat.

The orchestrator is the model the technician talks to. Other roles are delegated
to (cheap scout / grep / summarizer, expensive coder) so a long ticket does not
re-send every file read through the $50/hour model.

Roles are a fixed catalog. Member models are NOT restricted to the Models table
- any id the provider currently serves is allowed. A group's is_default flag
outranks the starred Models default for new chats.
"""

from __future__ import annotations

from types import SimpleNamespace
from typing import Any, Optional

from django.utils.text import slugify

ROLE_DEFINITIONS = {
    "orchestrator": (
        "You are the face of the chat. The technician talks only to you.\n"
        "Plan the work, decide which specialist to call, and synthesize their answers.\n"
        "Delegate recon, search, planning, review, and first-pass summaries so this "
        "conversation stays small. You keep the TRMM / ticket tools; specialists do not "
        "mutate devices. Do not dump raw logs or whole files into this thread if a "
        "specialist can compress them first. Prefer scout → planner → you apply the plan."
    ),
    "scout": (
        "Fast survey only. Look at a codebase, a ticket, a log, or a device and return a "
        "compressed briefing: what exists, what matters, what to ignore.\n"
        "Do not implement, do not draft a full plan, do not write code. Names, IDs, error "
        "strings, and file paths stay verbatim. Keep the answer short enough that the "
        "orchestrator can paste it forward without bloating the main chat."
    ),
    "files": (
        "Read and list files. Return the relevant contents, a tree, or the exact lines "
        "asked for. Do not edit, do not search the whole tree (that is grep), do not "
        "summarise away code the orchestrator asked to see. Quote paths exactly."
    ),
    "grep": (
        "Search the tree. Return matching paths and a short snippet per hit.\n"
        "Do not edit. Do not read entire files unless a single hit is useless without "
        "the surrounding function. Deduplicate. Prefer the tightest pattern that answers "
        "the question."
    ),
    "planner": (
        "Turn a goal plus whatever recon you are given into a concrete, ordered plan.\n"
        "Each step should say who does it (orchestrator / coder / operator / human) and "
        "what done looks like. Do not implement. Do not run tools that change anything. "
        "Call out risks, missing facts, and the smallest change that would work."
    ),
    "coder": (
        "Write or edit code. Touch only what the task requires.\n"
        "If a workspace is configured, make the change and summarise the diff. If not, "
        "return a complete patch the orchestrator can apply. Do not expand scope. Do not "
        "refactor adjacent code. Match the existing style."
    ),
    "operator": (
        "Draft the exact device fix: commands, scripts, expected output, and rollback.\n"
        "You do not run anything and you do not claim you did. The orchestrator keeps "
        "the TRMM tools and will execute. Be specific (full command lines, service names, "
        "paths). Flag anything disruptive (reboot, data loss, customer-facing)."
    ),
    "reviewer": (
        "Review a proposed change, script, or plan. Return: bugs, risks, missing steps, "
        "and a go / no-go.\n"
        "Do not rewrite it unless asked. Do not implement an alternative. Quote the "
        "line or step you are objecting to."
    ),
    "summarizer": (
        "Compress the given material into a briefing the orchestrator can work from.\n"
        "Keep names, IDs, numbers, error text, decisions, and open questions. Drop "
        "pleasantries, repeated tool noise, and anything the next turn does not need. "
        "This role exists so a long ticket stops re-sending millions of tokens."
    ),
}

ROLE_CATALOG = [
    {
        "id": "orchestrator",
        "label": "Orchestrator",
        "required": True,
        "kinds": ["coding", "it", "custom"],
        "description": ROLE_DEFINITIONS["orchestrator"],
    },
    {
        "id": "scout",
        "label": "Scout / recon",
        "required": False,
        "kinds": ["coding", "it", "custom"],
        "description": ROLE_DEFINITIONS["scout"],
    },
    {
        "id": "files",
        "label": "File browser",
        "required": False,
        "kinds": ["coding", "custom"],
        "description": ROLE_DEFINITIONS["files"],
    },
    {
        "id": "grep",
        "label": "Search",
        "required": False,
        "kinds": ["coding", "custom"],
        "description": ROLE_DEFINITIONS["grep"],
    },
    {
        "id": "planner",
        "label": "Planner",
        "required": False,
        "kinds": ["coding", "it", "custom"],
        "description": ROLE_DEFINITIONS["planner"],
    },
    {
        "id": "coder",
        "label": "Coder",
        "required": False,
        "kinds": ["coding", "custom"],
        "description": ROLE_DEFINITIONS["coder"],
    },
    {
        "id": "operator",
        "label": "Operator",
        "required": False,
        "kinds": ["it", "custom"],
        "description": ROLE_DEFINITIONS["operator"],
    },
    {
        "id": "reviewer",
        "label": "Reviewer",
        "required": False,
        "kinds": ["coding", "it", "custom"],
        "description": ROLE_DEFINITIONS["reviewer"],
    },
    {
        "id": "summarizer",
        "label": "Summarizer",
        "required": False,
        "kinds": ["coding", "it", "custom"],
        "description": ROLE_DEFINITIONS["summarizer"],
    },
]

ROLE_IDS = {r["id"] for r in ROLE_CATALOG}
REQUIRED_ROLES = {r["id"] for r in ROLE_CATALOG if r["required"]}

# Recommended rosters. Models are looked up against live providers at seed time;
# a member whose provider has no key is skipped, not fatal.
SEED_GROUPS = [
    {
        "name": "Coding",
        "slug": "coding",
        "kind": "coding",
        "is_default": False,
        "description": (
            "Multi-agent coding team. Sonnet orchestrates; Haiku does recon / files / "
            "grep / summaries; Grok 4.6 only gets called to write code. Keeps the expensive "
            "model's context small."
        ),
        "members": [
            ("orchestrator", "anthropic", "claude-sonnet-5", "Claude Sonnet 5", "medium"),
            ("scout", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
            ("files", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
            ("grep", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
            ("planner", "anthropic", "claude-sonnet-5", "Claude Sonnet 5", "medium"),
            ("coder", "xai", "grok-4.6", "grok-4.6", "high"),
            ("reviewer", "anthropic", "claude-sonnet-5", "Claude Sonnet 5", "medium"),
            ("summarizer", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
        ],
    },
    {
        "name": "Coding+",
        "slug": "coding-plus",
        "kind": "coding",
        "is_default": False,
        "description": (
            "Same team as Coding, but the coder is Claude Fable 5.1. Grok orchestrates "
            "and plans; Haiku does recon / files / grep / summaries; Sonnet reviews. Only "
            "the actual code-writing step pays for the top-tier model."
        ),
        "members": [
            ("orchestrator", "xai", "grok-4.6", "grok-4.6", "medium"),
            ("scout", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
            ("files", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
            ("grep", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
            ("planner", "xai", "grok-4.6", "grok-4.6", "medium"),
            ("coder", "anthropic", "claude-fable-5-1", "Claude Fable 5.1", "high"),
            ("reviewer", "anthropic", "claude-sonnet-5", "Claude Sonnet 5", "medium"),
            ("summarizer", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
        ],
    },
    {
        "name": "IT",
        "slug": "it",
        "kind": "it",
        "is_default": True,
        "description": (
            "Multi-agent IT / ticket team. Grok stays the face of the chat (same as "
            "today's default); Haiku scouts and summarises so a long ticket does not "
            "keep re-sending the whole transcript; Sonnet plans, drafts fixes, and reviews."
        ),
        "members": [
            ("orchestrator", "xai", "grok-4.6", "grok-4.6", "high"),
            ("scout", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
            ("planner", "anthropic", "claude-sonnet-5", "Claude Sonnet 5", "medium"),
            ("operator", "anthropic", "claude-sonnet-5", "Claude Sonnet 5", "medium"),
            ("reviewer", "anthropic", "claude-sonnet-5", "Claude Sonnet 5", "medium"),
            ("summarizer", "anthropic", "claude-haiku-4-5", "Claude Haiku 4.5", "low"),
        ],
    },
]


def public_groups():
    """Browser-safe list of enabled groups (no API keys)."""
    from core.models import AIAgentGroup

    out = []
    for g in AIAgentGroup.objects.filter(enabled=True).prefetch_related("members"):
        out.append(public_group(g))
    return out


def public_group(g) -> dict:
    orch = next((m for m in g.members.all() if m.role == "orchestrator" and m.enabled), None)
    return {
        "id": g.id,
        "name": g.name,
        "slug": g.slug,
        "kind": g.kind,
        "description": g.description,
        "is_default": g.is_default,
        "workspace": g.workspace or "",
        "orchestrator": (
            {
                "provider": orch.provider,
                "model_id": orch.model_id,
                "display_name": orch.display_name or orch.model_id,
                "thinking_level": orch.thinking_level,
            }
            if orch
            else None
        ),
        "roles": [
            {
                "role": m.role,
                "provider": m.provider,
                "model_id": m.model_id,
                "display_name": m.display_name or m.model_id,
                "thinking_level": m.thinking_level,
                "definition": m.definition or "",
            }
            for m in g.members.all()
            if m.enabled
        ],
    }


def blob_group(g) -> dict:
    """Full roster for the bridge (still no API keys - those live on providers)."""
    pub = public_group(g)
    pub["members"] = pub["roles"]
    return pub


def _provider_map():
    from core.models import AIProvider

    return {p.name: p for p in AIProvider.objects.filter(enabled=True)}


def resolve_orchestrator(group) -> Optional[SimpleNamespace]:
    """Return a chosen-model-shaped object for the group's orchestrator, or None."""
    member = next(
        (m for m in group.members.all() if m.role == "orchestrator" and m.enabled),
        None,
    )
    if not member:
        return None
    return resolve_member(member)


def resolve_member(member) -> Optional[SimpleNamespace]:
    providers = _provider_map()
    prov = providers.get(member.provider)
    if not prov or not prov.api_key:
        return None
    return SimpleNamespace(
        provider=prov,
        model_id=member.model_id,
        display_name=member.display_name or member.model_id,
        thinking_level=member.thinking_level or "medium",
    )


def pick_group(request) -> tuple[Optional[Any], list]:
    """Decide which group (if any) this new session should use.

    - group_id in the body (including null) is an explicit choice.
    - A resume without an explicit group_id does not apply the default
      (the existing session already has a model).
    - Otherwise the default group wins over the starred Models default.
    """
    from core.models import AIAgentGroup

    groups = list(AIAgentGroup.objects.filter(enabled=True).prefetch_related("members"))
    public = [public_group(g) for g in groups]
    by_id = {g.id: g for g in groups}

    data = getattr(request, "data", {}) or {}
    if "group_id" in data:
        raw = data.get("group_id")
        if raw in (None, "", 0, "0", "null"):
            return None, public
        try:
            gid = int(raw)
        except (TypeError, ValueError):
            return None, public
        return by_id.get(gid), public

    if data.get("resume_session"):
        return None, public

    default = next((g for g in groups if g.is_default), None)
    return default, public


def group_provider_keys(group_blob) -> dict:
    """API keys for every provider used by the group. Server-side only.
    Accepts either an AIAgentGroup or the public blob dict."""
    if not group_blob:
        return {}
    providers = _provider_map()
    names = set()
    if hasattr(group_blob, "members"):
        names = {m.provider for m in group_blob.members.all()}
    else:
        for m in (group_blob.get("members") or group_blob.get("roles") or []):
            if m.get("provider"):
                names.add(m["provider"])
        if group_blob.get("orchestrator", {}).get("provider"):
            names.add(group_blob["orchestrator"]["provider"])
    return {n: providers[n].api_key for n in names if n in providers and providers[n].api_key}


def apply_group(blob: dict, request, chosen):
    """Attach group metadata to the session blob. May replace `chosen` with the
    group's orchestrator. Returns the (possibly new) chosen model object."""
    group, public = pick_group(request)
    blob["agent_groups"] = public
    blob["agent_group"] = None
    blob["group_requested"] = "group_id" in (getattr(request, "data", None) or {})
    if not group:
        return chosen
    orch = resolve_orchestrator(group)
    if not orch:
        return chosen
    blob["agent_group"] = blob_group(group)
    return orch


def seed_builtin_groups(*, reset_members: bool = False) -> dict:
    """Create/update the Coding, Coding+ and IT groups. Idempotent.

    Members whose provider is not configured (or has no key) are skipped so a
    fresh box with only xAI still gets an IT group, just without the Anthropic
    specialists until that key is added.
    """
    from core.models import AIAgentGroup, AIAgentGroupMember, AIProvider

    have = {p.name for p in AIProvider.objects.filter(enabled=True) if p.api_key}
    created, updated, skipped = [], [], []
    for spec in SEED_GROUPS:
        group, was_created = AIAgentGroup.objects.update_or_create(
            slug=spec["slug"],
            defaults={
                "name": spec["name"],
                "kind": spec["kind"],
                "description": spec["description"],
                "enabled": True,
                "is_default": spec.get("is_default", False),
            },
        )
        (created if was_created else updated).append(spec["slug"])
        if reset_members or was_created:
            if reset_members:
                group.members.all().delete()
            for role, provider, model_id, display, thinking in spec["members"]:
                if provider not in have:
                    skipped.append(f"{spec['slug']}:{role} ({provider}/{model_id} — no key)")
                    continue
                AIAgentGroupMember.objects.update_or_create(
                    group=group,
                    role=role,
                    defaults={
                        "provider": provider,
                        "model_id": model_id,
                        "display_name": display,
                        "thinking_level": thinking,
                        "definition": ROLE_DEFINITIONS.get(role, ""),
                        "enabled": True,
                    },
                )
        # Re-assert a single default. The last seed spec with is_default wins.
        if spec.get("is_default"):
            AIAgentGroup.objects.exclude(pk=group.pk).update(is_default=False)
            if not group.is_default:
                group.is_default = True
                group.save(update_fields=["is_default"])
    return {"created": created, "updated": updated, "skipped": skipped}


def unique_slug(name: str, exclude_pk=None) -> str:
    base = slugify(name) or "group"
    slug = base
    from core.models import AIAgentGroup

    qs = AIAgentGroup.objects.all()
    if exclude_pk:
        qs = qs.exclude(pk=exclude_pk)
    n = 2
    while qs.filter(slug=slug).exists():
        slug = f"{base}-{n}"
        n += 1
    return slug
